src/Orleans.Runtime/Timers/GrainTimer.cs

```diff
@@ -176,22 +213,28 @@ public void Dispose()
         protected virtual void Dispose(bool disposing)
         {
             if (disposing)
+            {
                 DisposeTimer();
-            
-            asyncCallback = null;
+            }
         }
 
         private void DisposeTimer()
         {
-            var tmp = timer;
-            if (tmp == null) return;
+            var asyncTimer = Interlocked.CompareExchange(ref timer, null, timer);
```

Comment:
Rather than taking a lock during disposal, we can use an interlocked operation. This allows us to atomically set the timer to null so that we can dispose it. Note that the existing code sets the timer to null outside of a lock anyway: the new version of the code uses `timer is { IsStarted: true }` instead of `asyncCallback is null` to check if the timer has been disposed.